File: Graph/210. Course Schedule II/courseScheduleII.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
 var findOrder = function(numCourses, prerequisites) {
    let preMap = new Map();
    let visited = new Set();
    let cycle = new Set();
    let output = []
    
    for (let i = 0; i < numCourses; i++) {
        preMap.set(i, []);
    }
    for (let [course, pre] of prerequisites) {
        preMap.get(course).push(pre);
    }
    
    for (let i = 0; i < numCourses; i++) {
        if (dfs(i) === false) {
            return [];
        }
    }
    return output;
    
    function dfs(course) {
        if (cycle.has(course)) {
            return false;
        }
        if (visited.has(course)) {
            return true
        }
        
        cycle.add(course);
        for (let pre of preMap.get(course)) {
            if (dfs(pre) === false)
                return false;
        }
        
        cycle.delete(course);
        visited.add(course);
        output.push(course);
        return true; 
    }
};

File: Graph/210. Course Schedule II/courseScheduleII.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        # adjacency list of papers 
        pre_map = { c:[] for c in range(numCourses)}
        for [course, pre] in prerequisites:
            pre_map[course].append(pre)
            
        # a course has 3 possible states:
        # visited -> crs has been added to output
        # visiting -> crs has been added to output, but added to cycle
        # unvisited -> crs not added to output or cycle
        output = []
        visited, cycle = set(), set()
        
        def dfs(course):
            # detect cycle
            if course in cycle:
                return False
            if course in visited:
                return True
            
            cycle.add(course)
            # iterate pre in each course
            for pre in pre_map[course]:
                if not dfs(pre):
                    return False
            
            cycle.remove(course)
            visited.add(course)
            output.append(course)
            return True
        
        for course in range(numCourses):
            if not dfs(course):
                return []
        return output
            